### PR TITLE
Implement HTTPRouteFilterRequestHeaderModifier for HTTPRouteFilter

### PR DESCRIPTION
**Pull request by nak3**
This patch implements HTTPRouteFilterRequestHeaderModifier for HTTPRouteFilter.

[HTTPRouteFilter](https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.HTTPRouteFilter) is defined in `HTTPRouteForwardTo` and `HTTPRouteRule`. This patch
adds `HTTPRouteFilterRequestHeaderModifier` for both fields.

/cc @stevesloka @youngnick @sunjayBhatia

Signed-off-by: Kenjiro Nakayama <nakayamakenjiro@gmail.com>

### CODE REVIEW

**Review by stevesloka — CHANGES_REQUESTED** [review]
Overall this looks great @nak3, just thinking maybe we should try and serve up routes even if some headers are invalid vs failing quickly.

**Comment by nak3** [comment]
Thank you @stevesloka ! Updated.
Although I updated to stop invalidate the entire route by one invalid header, other existing functions like [headersPolicyRoute](https://github.com/projectcontour/contour/blob/c55d96a6419238a6b3574cd130caf29b9b7d0fff/internal/dag/policy.go#L136) and [headersPolicyService](https://github.com/projectcontour/contour/blob/c55d96a6419238a6b3574cd130caf29b9b7d0fff/internal/dag/policy.go#L93) seems to invalidate all the entire route/service by one invalid header. Is that alright?

File: internal/dag/gatewayapi_processor.go [hunk 1]
```diff
@@ -246,15 +247,53 @@ func (p *GatewayAPIProcessor) computeHTTPRoute(route *gatewayapi_v1alpha1.HTTPRo
...
 				continue
 			}
-			services = append(services, service)
+
+			var headerPolicy *HeadersPolicy
+			for _, filter := range forward.Filters {
+				switch filter.Type {
+				case gatewayapi_v1alpha1.HTTPRouteFilterRequestHeaderModifier:
+					if filter.RequestHeaderModifier.Add != nil {
+						routeAccessor.AddCondition(status.ConditionNotImplemented, metav1.ConditionTrue, status.ReasonHTTPRouteFilterType,
```
Comment (stevesloka) [inline internal/dag/gatewayapi_processor.go:256]: Sorry I missed this on the first pass. We should be able to handle "add". Internally it's just a "set", if the header key doesn't exist then Envoy should add the header.

File: internal/dag/policy.go [hunk 2]
```diff
@@ -188,6 +190,54 @@ func headersPolicyRoute(policy *contour_api_v1.HeadersPolicy, allowHostRewrite b
 	}, nil
 }
 
+// headersPolicyGatewayAPI builds a *HeaderPolicy for the supplied HTTPRequestHeaderFilter.
+// TODO: Take care about the order of operators once https://github.com/kubernetes-sigs/gateway-api/issues/480 was solved.
+func headersPolicyGatewayAPI(hf *gatewayapi_v1alpha1.HTTPRequestHeaderFilter) (*HeadersPolicy, error) {
```
Comment (stevesloka) [inline internal/dag/policy.go:195]: I think we should refactor the headersPolicy bits that exist today into something more generic such that we don't duplicate this logic. We could do that work on a separate PR, or refactor after this one.

**Review by stevesloka — CHANGES_REQUESTED** [review]
Sorry a few more things to think about @nak3.

**Comment by stevesloka** [comment]
//ref https://github.com/projectcontour/contour/issues/3559

File: internal/dag/gatewayapi_processor.go (hunk 1 above)
Comment (nak3) [reply internal/dag/gatewayapi_processor.go:280]: Umm... As per gateway-api spec, `add` appends the header even with the same key name.

ref: https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.HTTPRequestHeaderFilter

```
Input: GET /foo HTTP/1.1 my-header: foo
Config: add: {“my-header”: “bar”}
Output: GET /foo HTTP/1.1 my-header: foo my-header: bar
```

So we can handle the `add` operation when the header key does not exist, but we cannot handle it when the header key already exist as current `HeadersPolicy.Set` is `map[string]string`. I think the support of `add` operation can include the refactoring of headersPolicy.

File: internal/dag/policy.go (hunk 2 above)
Comment (nak3) [reply internal/dag/policy.go:195]: I think refactor after this PR by separate PR is better. I will open an issue ticket for it.

Comment (nak3) [reply internal/dag/policy.go:195]: Opened https://github.com/projectcontour/contour/issues/3562

File: internal/dag/gatewayapi_processor.go (hunk 1 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:280]: If the header already exists in the request Envoy will handle the adding bits. We have a setting that handles that: https://github.com/projectcontour/contour/blob/46d5ba9df6f6634143a534f8d0edd0d497cad2dc/internal/envoy/v3/route.go#L247

File: internal/dag/policy.go (hunk 2 above)
Comment (stevesloka) [reply internal/dag/policy.go:195]: 👍

File: _integration/testsuite/gatewayapi/006-rewritehost.yaml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /
+      filters:
+      - type: RequestHeaderModifier
+        requestHeaderModifier:
+          add:
+            Host: rewritten.com
```
Comment (nak3) [inline _integration/testsuite/gatewayapi/006-rewritehost.yaml:73]: Rewrite Host works here. But it does not work per-service. HTTPProxy also has the issue so I opened https://github.com/projectcontour/contour/issues/3565.

File: internal/dag/gatewayapi_processor.go (hunk 1 above)
Comment (nak3) [reply internal/dag/gatewayapi_processor.go]: I see! Updated. Than you!

**Review by stevesloka — APPROVED** [review]
/lgtm Thanks @nak3!!

**Review by sunjayBhatia — COMMENTED** [review]
Maybe some unit tests to ensure invalid headers etc. are rejected, headers are canonicalized, and duplicates are handled appropriately with status updates etc.?

Otherwise LGTM

**Comment by nak3** [comment]
@sunjayBhatia Sure, added invalid header test cases. Thank you.

**Review by sunjayBhatia — APPROVED** [review]

**Comment by nak3** [comment]
(Rebased to fix confliction.)
